### PR TITLE
win: fix returning thread id in uv_thread_self

### PR DESCRIPTION
**Pull request by saghul**
The main thread does not have the thread id TLS created.

Fixes: https://github.com/joyent/node/issues/25602

R=@piscisaureus, @orangemocha

### CODE REVIEW

**Comment by saghul** [comment]
Folks, I'm either totally off-base here, or I don't understand how this could previously work, before any thread was created by libuv, and even then. Also, since the test in test/test-thread-equal.c passes, (I don't know why, btw), I'm not sure I can write a test for this.

See the linked Node issue.

**Comment by orangemocha** [comment]
There were two changes that led to this behavior. 

The first one was https://github.com/libuv/libuv/commit/6591d000d13714c9e362942fbeaeca2b29507a25, which started using TLS for storing the thread HANDLE when the thread was created with uv_thread_create(). The TLS slot was declared with UV_THREAD_LOCAL (see https://github.com/libuv/libuv/blob/6591d000d13714c9e362942fbeaeca2b29507a25/src/win/thread.c#L120) which doesn't require any explicit allocation of the TLS slot. The main thread (or any thread not created with uv_thread_create) would have a uv__current_thread of NULL, which would be different from uv__current_thread for any thread created with uv_thread_create(). So far so good.

The second change was https://github.com/libuv/libuv/commit/a31024865b3146597cecc3d7f5bdf251da8b1a9b. The TLS slot is now allocated explicitly (https://github.com/libuv/libuv/blob/a31024865b3146597cecc3d7f5bdf251da8b1a9b/src/win/thread.c#L128-131 and https://github.com/libuv/libuv/blob/a31024865b3146597cecc3d7f5bdf251da8b1a9b/src/win/thread.c#L149) but we only doing this when creating the first thread with uv_thread_create(). If no threads are created that way, the slot is unallocated and calling uv_thread_self() from the main thread will fail.

File: src/win/thread.c [hunk 1]
```diff
@@ -191,7 +191,13 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
 
 
 uv_thread_t uv_thread_self(void) {
-  return (uv_thread_t) uv_key_get(&uv__current_thread_key);
+  void* self;
+  uv_once(&uv__current_thread_init_guard, uv__init_current_thread_key);
```
Comment (orangemocha) [inline src/win/thread.c]: Can we instead move this uv_once initialization to the uv_init() code path? That way we should be guarantee that the TLS slot is always allocated, and remove the uv_once checks everywhere else, which should have better performance.

File: src/win/thread.c [hunk 2]
```diff
@@ -191,7 +191,13 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+  self = uv_key_get(&uv__current_thread_key);
+  if (self == NULL)
+      /* This thread wasn't created with uv_thread_create. */
+      self = GetCurrentThread();
```
Comment (orangemocha) [inline src/win/thread.c]: Note that GetCurrentThread() always returns a pseudo-thread handle which is only valid in the context of the current thread, and if I remember correctly, the same for all threads. So returning NULL (which is the default value of the TLS slot) is just as good for the purpose of comparing uv_thread_t values returned by uv_thread_self().

https://msdn.microsoft.com/en-us/library/windows/desktop/ms683182(v=vs.85).aspx

In summary, I would recommend no change here, and move the uv__init_current_thread_key initialization to uv_init()

File: src/win/thread.c (hunk 1 above)
Comment (saghul) [inline src/win/thread.c]: Sounds reasonable, I'll update the code.

File: src/win/thread.c (hunk 2 above)
Comment (saghul) [inline src/win/thread.c]: Hum. The problem is that it would be NULL for any thread not created by libuv, not only the main thread. Right now, by accident or not, on Unix we return a valid thread id even if we didn't create the thread.

> Note that GetCurrentThread() always returns a pseudo-thread handle which is only valid in the context of the current thread, and if I remember correctly, the same for all threads.

Since we don't store it, wouldn't this suffice?

Comment (orangemocha) [inline src/win/thread.c]: Yes, it would be NULL for any thread not created by libuv (including the main thread). My point is that GetCurrentThread() returns a constant (I think it's -2), so NULL is just as good. 

In order to get a unique per-thread HANDLE for those threads that were not created by libuv, we would have to call DuplicateHandle on that pseudo-handle, but then we would also have to make sure we close those handles when the thread exits. AFAIK, no one has complained so far about libuv returning the same value for all threads not created by uv_thread_create().

Comment (saghul) [inline src/win/thread.c]: I see, thanks for clarifying. I''ll adjust this PR then.

**Comment by saghul** [comment]
@orangemocha updated, can you PTAL? I ended up leaving that uv_once there, it makes the change much simpler, because the key is static and I would need to move it to internal.h or something like that.

**Comment by orangemocha** [comment]
:+1:

**Comment by piscisaureus** [comment]
lgtm

**Comment by saghul** [comment]
Landed, thanks @orangemocha and @piscisaureus!
